### ISSUE TITLE
Use case: Modelling view with data from independent models

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#3105]
Consider this imaginary model:

```C#
[Database]
class Product {
  string Id;
  string Name;
  string Description;
}

[Database]
class ProductSupplier {
  int Rating;
}
```

Both classes have a relationship, but solving it is not trivial. Lets say we must do a set of selects to find a certain supplier we want. Bottomline: it's not warrant to expose suppliers from the `Product` class in terms of a property, not even one with a path (like `Product,Something.SomehingElse.Supplier`).

Now lets say I want to design a view model that support showing a grid where the first column is the product basics (Id, Name, Description), but the last column is the product supplier with the highest rating.

Id | Name | Description | Supplier rating
---- | ----- | ------ | ------
1 | Win10 | Windows 10 licence | 8

Whats my take here?

The most intuitive way would be to model a view that support an array of products.

**Products.json**
```json
{
  "Items" : [
    { "Id": 0, "Name" : "", "Description" : "", "SupplierRating": 0 }
  ]
}
```

And to use `Product` instances as my data:

```C#
var json = new Products();
json.Items = Db.SQL<Products>("SELECT ...");
```

But that would leave the rating empty.

One technically possible way would be do introduce some intermediate POCO, like `ProductWithSupplierRating`, that I feed with the product and in which I implement the complex finding of the highest supplier rating.

```C#
class ProductWithSupplierRating {
  Product product;
  int rating;

  public ProductWithSupplierRating(Product p) {
    product = p;
    FindRating();
  }

  int Rating { get { return rating; } }
  string Id { get { return product.Id; } }
  // [...]
}
```

And then

```C#
var json = new Products();
json.Items = Helper.SelectProductsAndCreateWithRatingList();
```

But that would mean I'd have to maintain more code that I want, and also make me pay a performance penalty and bloated memory.

Lets try find some best practice / pattern for this, as I think its a case we'll need to be prepared for.

/cc @jwillmer

**Comment by k-rus** [comment]
I don't fully understand the original schema. What is the relationship between `Product` and `ProductSuplier`? Is it many-to-many? Then an additional relationship class needs to be defined.

I believe setting rating property manually is an option. Or introducing code property, which retrieves the values.

**Comment by jwillmer** [comment]
In short the problem is that you like to display a table with property values from different classes that don't need to have a relationships (like product and country tax). To accumulate all the data from all sources you introduce a POCO (viewmodel) and like to use this object to send to the view.

**Comment by per-samuelsson** [comment]
> I believe setting rating property manually is an option. Or introducing code property, which retrieves the values.

Yes, of course, but it requires manual coding, and we dont like that. I'm trying to raise the question if we can come up with a smarter way to solve this general case.

**Comment by k-rus** [comment]
One possibility to write a query and then populate the result directly into json collection of objects. We discuss that such scenarios can be valuable and new SQL processor can support this by creating objects as result of the query. But this is future plans and details will need to be investigated further.

**Comment by per-samuelsson** [comment]
Adding Typed JSON label to this. I think that to come up with pattern will mostly involve extending that.

**Comment by per-samuelsson** [comment]
I considered this on and off for some time now, and also experimented a little.

An assumption: you would normally always have a "main" data type from the underlying model. In my above example, it would be `Product`. In @jwillmer's, it could be either `Product` or `ProductType`.

Based on that assumption, maybe we should promote the pattern to simply implement this kind of logic in the view model code behind. 

If we fix #2964, we could implement this like

**Product.json**
```json
{
  "Id": "",
  "Name": "",
  "Description": "",
  "SupplierRating": 0
}
```

**Product.json.cs**
```c#
partial class Product : Json {
  public int SupplierRating {
    get {
      return Helper.CalculateComplexProductSupplierRating();
    }
  }
}
```

I think that make a lot of sense, considering the tranditional role of the view model: to act like a middleman between the model and the view, and provide stuff to the view that neither fit the model nor the view.

**Comment by k-rus** [comment]
@per-samuelsson Interesting idea. It will be necessary to introduce naming convention to avoid confusion between model class and middleman class. Otherwise it might make difficult for bug hunting in the application code.

**Comment by Starcounter-Jack** [comment]
@per-samuelsson Have tou checked with @chrhol because binding i supposed to prioritize view model properties before viewmodel.Data properties. I thought your example would work already.
